jujugui/static/gui/src/app/jujulib/test-bakery.js

```diff
@@ -250,5 +261,133 @@ describe('Bakery', () => {
+    });
+
+    it('sets the content type correctly for the wait call', () => {
+      bakery._interact(
+        'http://example.com/visit', 'http://example.com/wait', ()=>{}, ()=>{});
+      assert.equal(client._sendRequest.callCount, 1);
+      assert.equal(
+        client._sendRequest.args[0][1], 'http://example.com/wait');
+      assert.equal(
+        client._sendRequest.args[0][2]['Content-Type'], 'application/json');
+    });
+
+    it('handles retry', () => {
+      client._sendRequest
+        .onFirstCall().callsArgWith(7, getResponse(true))
+        .onSecondCall().callsArgWith(7, getResponse(false));
+      bakery._interact(
+        'http://example.com/visit', 'http://example.com/wait', ()=>{}, ()=>{});
+      assert.equal(client._sendRequest.callCount, 2);
+    });
+
+    it('limits retries', () => {
+      client._sendRequest.callsArgWith(7, getResponse(true));
+      bakery._interact(
+        'http://example.com/visit', 'http://example.com/wait', ()=>{}, ()=>{});
+      assert.equal(client._sendRequest.callCount, 6); // 6 is retrycount limit
+    });
+
+    it('handles errors', () => {
+      client._sendRequest.callsArgWith(7, getResponse(true));
+      sinon.stub(bakery, '_getError').returns({'message': 'bad wolf'});
+      const ok = sinon.stub();
+      const fail = sinon.stub();
+      bakery._interact(
+        'http://example.com/visit', 'http://example.com/wait', ok, fail);
+      assert.equal(ok.callCount, 0);
+      assert.equal(fail.callCount, 1);
+      assert.equal(fail.args[0][0], 'cannot interact: bad wolf');
+    });
+
+    it('handles success', () => {
+      client._sendRequest.callsArgWith(7, getResponse(true));
+      sinon.stub(bakery, '_getError').returns(null);
+      const ok = sinon.stub();
+      const fail = sinon.stub();
+      bakery._interact(
+        'http://example.com/visit', 'http://example.com/wait', ok, fail);
+      assert.equal(ok.callCount, 1);
+      assert.equal(fail.callCount, 0);
+    });
+  });
+
+  describe('storage', () => {
+    it('exist', () => {
+      assert.isObject(storage);
+    });
+
+    it('sets items', () => {
+      storage.set('http://example.com/charmstore', 'foo', () => {});
+      assert.equal(fakeLocalStorage.store['charmstore'], 'foo');
+    });
+
+    it('gets items', () => {
+      fakeLocalStorage.store['charmstore'] = 'foo';
+      assert.equal('foo', storage.get('http://example.com/charmstore'));
+    });
+
+    describe('keys', () => {
+      it('special cases identity', () => {
+        // There's no identity url in services
+        Object.keys(storage._services).forEach(key => {
+          assert.notEqual('identity', key);
+        })
+        assert.equal(
+          'identity', storage._getKey('http://example.com/identity/'));
+      })
+      it('gets keys from services', () => {
+        assert.equal(
+          'charmstore', storage._getKey('http://example.com/charmstore'))
+      });
+    });
+
+    it('sets cookies for charmstore', () => {
```

Comment:
Actually it was supposed to be with the other storage tests. I've moved it into that describe.